### ISSUE TITLE
Default endpoint fallback doesn't work anymore

### ISSUE THREAD

**Issue by afbjorklund** [kubernetes-sigs/cri-tools#1041]
#### What happened:

As reported before, the default configuration is still broken for docker.

```console
$ sudo crictl --debug version
WARN[0000] runtime connect using default endpoints: [unix:///var/run/dockershim.sock unix:///run/containerd/containerd.sock unix:///run/crio/crio.sock unix:///var/run/cri-dockerd.sock]. As the default settings are now deprecated, you should set the endpoint instead. 
ERRO[0000] unable to determine runtime API version: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial unix /var/run/dockershim.sock: connect: no such file or directory" 
E1212 19:23:01.893406   62589 remote_runtime.go:189] "Version from runtime service failed" err="rpc error: code = Unimplemented desc = unknown service runtime.v1alpha2.RuntimeService"
FATA[0000] getting the runtime version: rpc error: code = Unimplemented desc = unknown service runtime.v1alpha2.RuntimeService 
```

#### What you expected to happen:

Since it still has a list of default endpoints, you would expect them to work.

```console
$ sudo crictl version
Version:  0.1.0
RuntimeName:  docker
RuntimeVersion:  20.10.21
RuntimeApiVersion:  1.41.0
```

#### How to reproduce it (as minimally and precisely as possible):

Install crictl and cri-dockerd with the default configuration (i.e. none)

Note that the docker configuration of containerd is **not** the default:

#### Anything else we need to know?:

This happens because docker has been moved _after_ containerd, but containerd doesn't have the cri plugin enabled.

```
disabled_plugins = ["cri"]
```

It worked with dockershim, since it was before containerd. But it fails with cri-dockerd, since it comes after containerd.

The recommended workaround is to set up `/etc/crictl.yaml` properly, and not rely on the deprecated cri-tools defaults.

```
runtime-endpoint: unix:///var/run/cri-dockerd.sock
```

#### Environment:

- Container runtime or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Others: crictl version v1.25.0

**Comment by afbjorklund** [comment]
Previous discussion:

* https://github.com/kubernetes-sigs/cri-tools/issues/893#issuecomment-1259866272

**Comment by afbjorklund** [comment]
Actually, it seems to be failing also for containerd. Something about error now being fatal ?

v1.25.0

```
DEBU[0000] get runtime connection                       
WARN[0000] runtime connect using default endpoints: [unix:///var/run/dockershim.sock unix:///run/containerd/containerd.sock unix:///run/crio/crio.sock unix:///var/run/cri-dockerd.sock]. As the default settings are now deprecated, you should set the endpoint instead. 
DEBU[0000] Note that performance maybe affected as each default connection attempt takes n-seconds to complete before timing out and going to the next in sequence. 
DEBU[0000] Connect using endpoint "unix:///var/run/dockershim.sock" with "2s" timeout 
ERRO[0000] unable to determine runtime API version: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial unix /var/run/dockershim.sock: connect: no such file or directory" 
DEBU[0000] Connect using endpoint "unix:///run/containerd/containerd.sock" with "2s" timeout 
DEBU[0000] Connected successfully using endpoint: unix:///run/containerd/containerd.sock 
```

v1.26.0

```
DEBU[0000] get runtime connection                       
WARN[0000] runtime connect using default endpoints: [unix:///var/run/dockershim.sock unix:///run/containerd/containerd.sock unix:///run/crio/crio.sock unix:///var/run/cri-dockerd.sock]. As the default settings are now deprecated, you should set the endpoint instead. 
DEBU[0000] Note that performance maybe affected as each default connection attempt takes n-seconds to complete before timing out and going to the next in sequence. 
DEBU[0000] Connect using endpoint "unix:///var/run/dockershim.sock" with "2s" timeout 
DEBU[0000] Connected successfully using endpoint: unix:///var/run/dockershim.sock 
DEBU[0000] VersionRequest: &VersionRequest{Version:v1,} 
E1213 20:22:57.677536   53350 remote_runtime.go:145] "Version from runtime service failed" err="rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing dial unix /var/run/dockershim.sock: connect: no such file or directory\""
DEBU[0000] VersionResponse: nil                         
FATA[0000] getting the runtime version: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial unix /var/run/dockershim.sock: connect: no such file or directory" 
```

**Comment by afbjorklund** [comment]
It turns out that it wasn't checking the errors, and thus only the first endpoint was ever used...

When the error is handled, then both containerd and dockerd work (i.e. if they support V1)

```
WARN[0000] runtime connect using default endpoints: [unix:///var/run/dockershim.sock unix:///run/containerd/containerd.sock unix:///run/crio/crio.sock unix:///var/run/cri-dockerd.sock]. As the default settings are now deprecated, you should set the endpoint instead. 
ERRO[0000] validate service connection: CRI v1 runtime API is not available for endpoint "unix:///var/run/dockershim.sock": rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial unix /var/run/dockershim.sock: connect: no such file or directory" 
ERRO[0000] validate service connection: CRI v1 runtime API is not implemented for endpoint "unix:///run/containerd/containerd.sock": rpc error: code = Unimplemented desc = unknown service runtime.v1.RuntimeService 
ERRO[0000] validate service connection: CRI v1 runtime API is not available for endpoint "unix:///run/crio/crio.sock": rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial unix /run/crio/crio.sock: connect: no such file or directory" 
ERRO[0000] validate service connection: CRI v1 runtime API is not implemented for endpoint "unix:///var/run/cri-dockerd.sock": rpc error: code = Unimplemented desc = unknown service runtime.v1.RuntimeService 
FATA[0000] validate service connection: CRI v1 runtime API is not implemented for endpoint "unix:///var/run/cri-dockerd.sock": rpc error: code = Unimplemented desc = unknown service runtime.v1.RuntimeService 
```

```go
--- a/vendor/k8s.io/kubernetes/pkg/kubelet/cri/remote/remote_runtime.go
+++ b/vendor/k8s.io/kubernetes/pkg/kubelet/cri/remote/remote_runtime.go
@@ -122,6 +122,8 @@ func (r *remoteRuntimeService) validateServiceConnection(ctx context.Context, co
 
        } else if status.Code(err) == codes.Unimplemented {
                return fmt.Errorf("CRI v1 runtime API is not implemented for endpoint %q: %w", endpoint, err)
+       } else if status.Code(err) == codes.Unavailable {
+               return fmt.Errorf("CRI v1 runtime API is not available for endpoint %q: %w", endpoint, err)
        }
 
        return nil
```

**Comment by saschagrunert** [comment]
@afbjorklund I agree we can fix that on the Kubernetes side, thank you for the detailed investigation!

**Comment by afbjorklund** [comment]
I couldn't find any test cases at all for `crictl`, it seems like only `critest` is being regression tested ?

**Comment by afbjorklund** [comment]
Fixed on master now, but not in the bundled version (1.26.0)

```console
$ sudo crictl version
WARN[0000] runtime connect using default endpoints: [unix:///var/run/dockershim.sock unix:///run/containerd/containerd.sock unix:///run/crio/crio.sock unix:///var/run/cri-dockerd.sock]. As the default settings are now deprecated, you should set the endpoint instead. 
ERRO[0000] validate service connection: validate CRI v1 runtime API for endpoint "unix:///var/run/dockershim.sock": rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial unix /var/run/dockershim.sock: connect: no such file or directory" 
ERRO[0000] validate service connection: validate CRI v1 runtime API for endpoint "unix:///run/containerd/containerd.sock": rpc error: code = Unimplemented desc = unknown service runtime.v1.RuntimeService 
ERRO[0000] validate service connection: validate CRI v1 runtime API for endpoint "unix:///run/crio/crio.sock": rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial unix /run/crio/crio.sock: connect: no such file or directory" 
Version:  0.1.0
RuntimeName:  docker
RuntimeVersion:  23.0.3
RuntimeApiVersion:  v1

```

**Comment by afbjorklund** [comment]
I guess this will be fixed with kubernetes 1.28, when cri-tools is bumped to 1.27

**Comment by rajeeshckr** [comment]
Thanks for raising this issue. This failed for us while upgrading to k8s 1.26. Having the `/etc/crictl.yaml` pointing to the `unix:///run/containerd/containerd.sock` fixed the issue.

```
crictl pull [registry.k8s.io/kube-apiserver:v1.26.7](http://registry.k8s.io/kube-apiserver:v1.26.7)
WARN[0000] image connect using default endpoints: [unix:///var/run/dockershim.sock unix:///run/containerd/containerd.sock unix:///run/crio/crio.sock unix:///var/run/cri-dockerd.sock]. As the default settings are now deprecated, you should set the endpoint instead.
E0802 02:08:26.677024   58257 remote_image.go:171] "PullImage from image service failed" err="rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing dial unix /var/run/dockershim.sock: connect: no such file or directory\"" image="[registry.k8s.io/kube-apiserver:v1.26.7](http://registry.k8s.io/kube-apiserver:v1.26.7)"
FATA[0000] pulling image: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial unix /var/run/dockershim.sock: connect: no such file or directory"
```

**Comment by afbjorklund** [comment]
Fixed in kubernetes `v1.28.0-alpha.1`, by upgrading to cri-tools `v1.27.0`

https://github.com/kubernetes/kubernetes/commit/bc00f5c071ff34b0122621d826a34619bea1f2f7

Still looks for dockershim first, but other than that it seems to be working.

* https://github.com/kubernetes-sigs/cri-tools/issues/870
